File: leetcode/algorithms/118_pascals_triangle/main.py
from typing import List


class PascalsTriangle:
    """
    # DP
    # 
    # Complexities:
    #   N - `numRows`
    #   - Time Complexity: O(N²)
    #   - Space Complexity: O(N²)
    """
    def generate(self, numRows: int) -> List[List[int]]:
        result = []

        for i in range(0, numRows):
            row = []
            for j in range(0, i + 1):
                if j == 0 or j == i:
                    row.append(1)
                else:
                    row.append(result[i - 1][j - 1] + result[i - 1][j])

            result.append(row)

        return result


    # Solution
    """
    # Solution 1
    #
    # DP
    # 
    # Complexities:
    #   N - `numRows`
    #   - Time Complexity: O(N²)
    #   - Space Complexity: O(N²)
    """
    def solution1(self, numRows: int) -> List[List[int]]:
        triangle = []

        for i in range(numRows):
            row = [1] * (i + 1)

            for j in range(1, i):
                row[j] = triangle[i-1][j-1] + triangle[i-1][j]

            triangle.append(row)

        return triangle

    """
    # Solution 2
    #
    # Pythonic Way - zip
    # 
    # Complexities:
    #   N - `numRows`
    #   - Time Complexity: O(N²)
    #   - Space Complexity: O(N²)
    """
    def solution2(self, numRows: int) -> List[List[int]]:
        triangle = [[1]]

        for _ in range(numRows - 1):
            prev_row = triangle[-1]

            new_row = [a + b for a, b in zip([0] + prev_row, prev_row + [0])]

            triangle.append(new_row)

        return triangle

File: leetcode/algorithms/118_pascals_triangle/main.js
/**
 * DP
 * 
 * Complexities:
 *   N - `numRows`
 *   - Time Complexity: O(N²)
 *   - Space Complexity: O(N²)
 */
/**
 * @param {number} numRows
 * @return {number[][]}
 */
var generate = function (numRows) {
  const result = [];
  for (let i = 0; i < numRows; i++) {
    const row = [];
    for (let j = 0; j <= i; j++) {
      if (j === 0 || j === i) {
        row.push(1);
      } else {
        row.push(result[i - 1][j - 1] + result[i - 1][j]);
      }
    }
    result.push(row);
  }
  return result;
};


// Solution
/**
 * DP
 * 
 * Complexities:
 *   N - `numRows`
 *   - Time Complexity: O(N²)
 *   - Space Complexity: O(N²)
 */
/**
 * @param {number} numRows
 * @return {number[][]}
 */
var solution = function (numRows) {
  if (numRows === 0) {
    return [];
  }

  const triangle = [[1]];

  for (let i = 1; i < numRows; i++) {
    const prevRow = triangle[i - 1];

    const currentRow = new Array(i + 1).fill(1);

    for (let j = 1; j < i; j++) {
      currentRow[j] = prevRow[j - 1] + prevRow[j];
    }

    triangle.push(currentRow);
  }

  return triangle;
};
